commit b4d14632bd9e00e2edb11e2da358cdf4d1cf47cd
commit ab616fbb6d95c450ec7d248a00fffcd90e32a71a
Author: GeorgeClark <GeorgeClark@0b4bb1d4-4e5a-0410-9cc4-b2b747904278>
Date:   Mon Dec 10 00:49:20 2012 +0000

    Item12285: Better validations
    
    git-svn-id: http://svn.foswiki.org/trunk@16178 0b4bb1d4-4e5a-0410-9cc4-b2b747904278

~~~PATCH 72c86f0c71519caf6d26efbe174739f6  lib/Foswiki/Macros/MAKETEXT.pm (Foswiki 1.1.0 - 1.1.2)
--- lib/Foswiki/Macros/MAKETEXT.pm.orig	2012-12-09 20:20:07.865197810 -0500
+++ lib/Foswiki/Macros/MAKETEXT.pm	2012-12-09 23:50:00.655040822 -0500
@@ -4,6 +4,15 @@
 use strict;
 use warnings;
 
+use Locale::Maketext;
+my $escape =
+  ( $Foswiki::cfg{UserInterfaceInternationalisation}
+      && ( Locale::Maketext->VERSION() < 1.23 ) );
+
+my $max;
+my $min;
+my $param_error;
+
 sub MAKETEXT {
     my ( $this, $params ) = @_;
 
@@ -18,11 +27,16 @@
     $str =~ s/~~\[/~[/g;
     $str =~ s/~~\]/~]/g;
 
+    $max         = 0;
+    $min         = 1;
+    $param_error = 0;
+
     # unescape parameters and calculate highest parameter number:
-    my $max = 0;
-    $str =~ s/~\[(\_(\d+))~\]/ $max = $2 if ($2 > $max); "[$1]"/ge;
-    $str =~
-s/~\[(\*,\_(\d+),[^,]+(,([^,]+))?)~\]/ $max = $2 if ($2 > $max); "[$1]"/ge;
+    $str =~ s/~\[(\_(\d+))~\]/_validate($1, $2)/ge;
+    $str =~ s/~\[(\*,\_(\d+),[^,]+(,([^,]+))?)~\]/ _validate($1, $2)/ge;
+    return $str if ($param_error);
+
+    $str =~ s#\\#\\\\#g if $escape;
 
     # get the args to be interpolated.
     my $argsStr = $params->{args} || "";
@@ -47,6 +61,22 @@
     return $result;
 }
 
+sub _validate {
+    $max = $_[1] if ( $_[1] > $max );
+    $min = $_[1] if ( $_[1] < $min );
+    if ( $_[1] > 100 ) {
+        $param_error = 1;
+        return
+"<span class=\"foswikiAlert\">Excessive parameter number $max, MAKETEXT rejected.</span>";
+    }
+    if ( $_[1] < 1 ) {
+        $param_error = 1;
+        return
+"<span class=\"foswikiAlert\">Invalid parameter <code>\"$_[0]\"</code>, MAKETEXT rejected.</span>";
+    }
+    return "[$_[0]]";
+}
+
 1;
 __END__
 Foswiki - The Free and Open Source Wiki, http://foswiki.org/

~~~PATCH  160f04fc478c5f9b81d2ef6c9e614074  lib/Foswiki/Macros/MAKETEXT.pm (Foswiki 1.1.3 - 1.1.6)
--- lib/Foswiki/Macros/MAKETEXT.pm.orig	2012-12-09 20:23:52.457636561 -0500
+++ lib/Foswiki/Macros/MAKETEXT.pm	2012-12-09 20:24:15.624372152 -0500
@@ -4,6 +4,10 @@
 use strict;
 use warnings;
 
+use Locale::Maketext;
+my $escape =
+  ( $Foswiki::cfg{UserInterfaceInternationalisation}
+      && ( Locale::Maketext->VERSION() < 1.23 ) );
+
+my $max;
+my $min;
+my $param_error;
+
 sub MAKETEXT {
     my ( $this, $params ) = @_;
 
@@ -18,11 +22,16 @@
     $str =~ s/~~\[/~[/g;
     $str =~ s/~~\]/~]/g;
 
+    $max         = 0;
+    $min         = 1;
+    $param_error = 0;
+
     # unescape parameters and calculate highest parameter number:
-    my $max = 0;
-    $str =~ s/~\[(\_(\d+))~\]/ $max = $2 if ($2 > $max); "[$1]"/ge;
-    $str =~
-s/~\[(\*,\_(\d+),[^,]+(,([^,]+))?)~\]/ $max = $2 if ($2 > $max); "[$1]"/ge;
+    $str =~ s/~\[(\_(\d+))~\]/_validate($1, $2)/ge;
+    $str =~ s/~\[(\*,\_(\d+),[^,]+(,([^,]+))?)~\]/ _validate($1, $2)/ge;
+    return $str if ($param_error);
+
+    $str =~ s#\\#\\\\#g if $escape;
 
     # get the args to be interpolated.
     my $argsStr = $params->{args} || "";
@@ -47,6 +56,22 @@
     return $result;
 }
 
+sub _validate {
+    $max = $_[1] if ( $_[1] > $max );
+    $min = $_[1] if ( $_[1] < $min );
+    if ( $_[1] > 100 ) {
+        $param_error = 1;
+        return
+"<span class=\"foswikiAlert\">Excessive parameter number $max, MAKETEXT rejected.</span>";
+    }
+    if ( $_[1] < 1 ) {
+        $param_error = 1;
+        return
+"<span class=\"foswikiAlert\">Invalid parameter <code>\"$_[0]\"</code>, MAKETEXT rejected.</span>";
+    }
+    return "[$_[0]]";
+}
+
 1;
 __END__
 Foswiki - The Free and Open Source Wiki, http://foswiki.org/


